File: src/math-engine.js
/**
 * Matematikai primitívek a 0-1 Roulette logikához
 * Port: mathEngine.ts → vanilla JS
 */

const GOLDEN_RATIO = (Math.sqrt(5) - 1) / 2; // ~0.6180339887
const PLASTIC_RATIO = 1.324717957244746;

function normalizeAngle(angle) {
  const mod = angle % 360;
  return mod < 0 ? mod + 360 : mod;
}

/**
 * Determinisztikus irányzék szög számítása N. lépésnél.
 * Ugyanolyan input → mindig ugyanolyan kimenet.
 */
function getReticleAngle(step, engineType, matrixEnergy) {
  const normalizedEnergy = matrixEnergy / 100;

  switch (engineType) {
    case 'weyl': {
      const fraction = (step * GOLDEN_RATIO) % 1;
      return fraction * 360;
    }
    case 'arnold': {
      const omega = GOLDEN_RATIO;
      const k = 0.5 + 2.0 * normalizedEnergy;
      let theta = 0.231;
      for (let i = 0; i < step; i++) {
        theta = (theta + omega - (k / (2 * Math.PI)) * Math.sin(2 * Math.PI * theta)) % 1;
        if (theta < 0) theta += 1;
      }
      return theta * 360;
    }
    case 'chebyshev': {
      const ratio = 1 / PLASTIC_RATIO;
      const val = Math.abs(Math.cos(step * ratio * Math.PI + normalizedEnergy * 2));
      return val * 360;
    }
    default:
      return (step * 45) % 360;
  }
}

/**
 * A lufi (0-forma) sugara adott polár-szögnél.
 * Rugalmas mag: feszültség, térfogat-megőrzés, behúzódás a reticle irányából.
 */
function getBalloonRadiusAtAngle(theta, baseRadius, reticleAngleDeg, tension, resonance, step) {
  const reticleAngleRad = (reticleAngleDeg * Math.PI) / 180;

  const harmonicCount = 4 + Math.round(resonance * 8);
  const waveAmple = 0.08 * (1 - tension);
  const breathing = Math.sin(step * 0.05) * 0.02;
  const naturalShape = baseRadius * (1 + waveAmple * Math.sin(harmonicCount * theta + step * 0.02) + breathing);

  const angleDiff = Math.atan2(Math.sin(theta - reticleAngleRad), Math.cos(theta - reticleAngleRad));
  const indentWidth = 0.45;
  const indentDepth = 0.22 * (1 - tension);
  const localDent = indentDepth * Math.exp(-(angleDiff * angleDiff) / (2 * indentWidth * indentWidth));

  const bulgeAngle1 = reticleAngleRad + Math.PI;
  const bulgeAngle2 = reticleAngleRad + Math.PI / 2;
  const bulgeAngle3 = reticleAngleRad - Math.PI / 2;

  const diff1 = Math.atan2(Math.sin(theta - bulgeAngle1), Math.cos(theta - bulgeAngle1));
  const diff2 = Math.atan2(Math.sin(theta - bulgeAngle2), Math.cos(theta - bulgeAngle2));
  const diff3 = Math.atan2(Math.sin(theta - bulgeAngle3), Math.cos(theta - bulgeAngle3));

  const bulgeWidth = 0.6;
  const bulgeValue = (indentDepth / 2) * (
    0.5 * Math.exp(-(diff1 * diff1) / (2 * bulgeWidth * bulgeWidth)) +
    0.25 * Math.exp(-(diff2 * diff2) / (2 * bulgeWidth * bulgeWidth)) +
    0.25 * Math.exp(-(diff3 * diff3) / (2 * bulgeWidth * bulgeWidth))
  );

  return Math.max(0.08, naturalShape - localDent + bulgeValue);
}

/**
 * Kardinális erőeloszlás számítása (É/K/D/Ny %).
 */
function calculateCardinalDistribution(reticleAngle, baseRadius, tension, resonance, step) {
  const reticleRad = (reticleAngle * Math.PI) / 180;

  const nRadius = getBalloonRadiusAtAngle(0, baseRadius, reticleAngle, tension, resonance, step);
  const eRadius = getBalloonRadiusAtAngle(Math.PI / 2, baseRadius, reticleAngle, tension, resonance, step);
  const sRadius = getBalloonRadiusAtAngle(Math.PI, baseRadius, reticleAngle, tension, resonance, step);
  const wRadius = getBalloonRadiusAtAngle((3 * Math.PI) / 2, baseRadius, reticleAngle, tension, resonance, step);

  const nWeight = Math.max(0, Math.cos(reticleRad - 0));
  const eWeight = Math.max(0, Math.cos(reticleRad - Math.PI / 2));
  const sWeight = Math.max(0, Math.cos(reticleRad - Math.PI));
  const wWeight = Math.max(0, Math.cos(reticleRad - (3 * Math.PI) / 2));

  let rawN = Math.max(1, nRadius * (1.0 + nWeight * 2.0) * 10);
  let rawE = Math.max(1, eRadius * (1.0 + eWeight * 2.0) * 10);
  let rawS = Math.max(1, sRadius * (1.0 + sWeight * 2.0) * 10);
  let rawW = Math.max(1, wRadius * (1.0 + wWeight * 2.0) * 10);

  const total = rawN + rawE + rawS + rawW;

  const north = Math.round((rawN / total) * 100);
  const east  = Math.round((rawE / total) * 100);
  const south = Math.round((rawS / total) * 100);
  const west  = Math.max(0, 100 - north - east - south);

  const forces = { north, east, south, west };
  let dominant = 'north';
  let maxVal = -1;
  for (const [dir, val] of Object.entries(forces)) {
    if (val > maxVal) { maxVal = val; dominant = dir; }
  }

  return { north, east, south, west, dominant };
}

/**
 * Mutató rezgés számítása – hang sebessége alapján (343 m/s)
 *
 * A mutató kar fizikai analógiában: az irányzék szög változása → ívhossz → rezgési frekvencia.
 * Kettős skála: fény (nm) a rétegekhez, hang (Hz) a mutatóhoz.
 *
 * @param {number} prevAngleDeg   előző szög (fok)
 * @param {number} currAngleDeg   jelenlegi szög (fok)
 * @param {number} armLength      kar hossza méterbem (alapértelmezett: 0.07 m = 7cm)
 * @returns {{ freqHz: number, ampNorm: number, arcM: number, wavelengthM: number, note: string }}
 */
function calcPointerVibration(prevAngleDeg, currAngleDeg, armLength = 0.07) {
  const V_SOUND = 343; // m/s

  let delta = Math.abs(currAngleDeg - prevAngleDeg);
  if (delta > 180) delta = 360 - delta; // legrövidebb ív

  const deltaRad   = delta * Math.PI / 180;
  const arcM       = armLength * deltaRad;          // ívhossz (m)
  const ampNorm    = Math.min(1.0, delta / 180);    // 0-1 amplitúdó

  // f = v / λ, félhullám-rezonancia: λ = 2 × ívhossz
  const freqHz = arcM > 0.0001 ? V_SOUND / (2 * arcM) : 0;

  // Hang hangjegy közelítés
  const NOTE_NAMES = ['C','C#','D','D#','E','F','F#','G','G#','A','A#','B'];
  let note = '–';
  if (freqHz > 20 && freqHz < 20000) {
    const midi = Math.round(69 + 12 * Math.log2(freqHz / 440));
    note = NOTE_NAMES[((midi % 12) + 12) % 12] + Math.floor(midi / 12 - 1);
  } else if (freqHz >= 20000) {
    note = 'ultraszound';
  } else {
    note = 'infraszound';
  }

  return {
    freqHz:     Math.round(freqHz),
    ampNorm,
    arcM:       +arcM.toFixed(5),
    wavelengthM: arcM > 0 ? +(2 * arcM).toFixed(5) : 0,
    note
  };
}

/**
 * Dinamikus alapátmérő számítása – sávos/aberrált módban.
 *
 * Módok:
 *   static     – rögzített átmérő (default 0.48)
 *   dynamic    – szinuszos oszcilláció lépésenként
 *   aberrated  – nem-lineáris, több-harmonikus tágulás (korlátlan mozgás)
 *
 * Az aberrált módban a burok különböző szögű sávjai különböző fázisban tágulnak,
 * ezért az átmérő polárszögtől is függ → true aberráció.
 */
function getDynamicBaseRadius(step, matrixEnergy, mode = 'static', theta = 0) {
  const e = matrixEnergy / 100;
  const t = step / 80; // normalizált idő

  if (mode === 'static') return 0.48;

  if (mode === 'dynamic') {
    return 0.32 + 0.20 * Math.abs(Math.sin(t * Math.PI + e * Math.PI));
  }

  if (mode === 'aberrated') {
    // Polárszög-függő sávos tágulás – különböző "fény-sávok" eltérő fázisban
    const bandPhase = Math.sin(4 * theta) * 0.5 + 0.5; // 0-1, szögenként változó
    return 0.28
      + 0.14 * Math.abs(Math.sin(t * Math.PI * GOLDEN_RATIO + theta))
      + 0.10 * Math.abs(Math.sin(t * Math.PI * PLASTIC_RATIO + e * Math.PI))
      + 0.06 * bandPhase * Math.abs(Math.sin(t * 2 * Math.PI + 4 * theta));
  }

  return 0.48;
}

/**
 * Aberrált sávos ballonsugar – az alap átmérő is dinamikus és szögfüggő.
 */
function getBalloonRadiusAberrated(theta, step, matrixEnergy, reticleAngleDeg, tension, resonance, mode = 'static') {
  const dynamicBase = getDynamicBaseRadius(step, matrixEnergy, mode, theta);
  return getBalloonRadiusAtAngle(theta, dynamicBase, reticleAngleDeg, tension, resonance, step);
}

export {
  GOLDEN_RATIO, PLASTIC_RATIO,
  normalizeAngle, getReticleAngle,
  getBalloonRadiusAtAngle, calculateCardinalDistribution,
  calcPointerVibration, getDynamicBaseRadius, getBalloonRadiusAberrated
};

File: java/src/main/java/freetranslator/MathEngine.java
package freetranslator;

/**
 * FreeTranslator v2.1 – Math Engine (Java port)
 *
 * Determinisztikus reticle szög generátorok:
 *   Weyl     – arany arány alapú egyenletes lefedés
 *   Arnold   – Arnold macska térkép (kaotikus, de periodikus)
 *   Chebyshev – Chebyshev-polinóm (ergodikus)
 *
 * Burok (0-forma) sugár: rugalmas elliptikus deformáció
 * Rezgés:  hang sebessége v = 343 m/s referencia
 * Dinamikus sáv: static / dynamic / aberrated módok
 */
public class MathEngine {

    public static final double PHI    = (1.0 + Math.sqrt(5)) / 2.0;
    public static final double SOUND  = 343.0;  // m/s
    public static final double ARM    = 0.07;   // reticle kar hossz (m)

    // ── Reticle szög generátorok ─────────────────────────────────────────────

    /** Weyl-sorozat: szög = frac(step × φ) × 360 */
    public static double weylAngle(int step) {
        return frac(step * PHI) * 360.0;
    }

    /** Arnold-macska térkép: x_n+1 = (x_n + y_n) mod 1; y_n+1 = (x_n + 2y_n) mod 1 */
    public static double arnoldAngle(int step) {
        double x = 0.5, y = 0.5;
        for (int i = 0; i < step % 999 + 1; i++) {
            double nx = frac(x + y);
            double ny = frac(x + 2 * y);
            x = nx; y = ny;
        }
        return x * 360.0;
    }

    /** Chebyshev-iteráció: x_n+1 = cos(arccos(x_n) × 2) */
    public static double chebyshevAngle(int step) {
        double x = 0.7;
        for (int i = 0; i < step % 999 + 1; i++) {
            x = Math.cos(2.0 * Math.acos(Math.max(-1, Math.min(1, x))));
        }
        return ((x + 1.0) / 2.0) * 360.0;
    }

    /**
     * Reticle szög kiválasztása motor típus szerint.
     * @param step         lépésszám
     * @param engineType   "weyl" | "arnold" | "chebyshev"
     * @param matrixEnergy 0–100 energia érték (perturbáció hozzáadása)
     */
    public static double getReticleAngle(int step, String engineType, double matrixEnergy) {
        double base;
        switch (engineType == null ? "weyl" : engineType) {
            case "arnold":    base = arnoldAngle(step);    break;
            case "chebyshev": base = chebyshevAngle(step); break;
            default:          base = weylAngle(step);      break;
        }
        double perturbation = (matrixEnergy / 100.0) * Math.sin(step * 0.314) * 15.0;
        return normalizeAngle(base + perturbation);
    }

    // ── Burok (0-forma) sugár ────────────────────────────────────────────────

    /**
     * Rugalmas burok sugár adott szögnél.
     * Szinusz-alapú deformáció a reticle szög irányára merőlegesen.
     */
    public static double getBalloonRadius(double thetaDeg, double baseR,
                                          double retAngleDeg, double tension,
                                          double resonance, int step) {
        double theta = Math.toRadians(thetaDeg);
        double retR  = Math.toRadians(retAngleDeg);
        double diff  = theta - retR;

        double deform = (1.0 - tension) * resonance * 0.25 * Math.cos(2.0 * diff);
        double wave   = 0.04 * Math.sin(3.0 * theta + step * 0.05);
        return baseR * (1.0 + deform + wave);
    }

    /** Statikus alapsugár (tension alapú) */
    public static double staticBase(double tension) {
        return 0.35 + tension * 0.25;
    }

    /** Dinamikus alapsugár – szinuszos oszcilláció */
    public static double dynamicBase(int step, double matrixEnergy) {
        return 0.45 + 0.12 * Math.sin(step * 0.08) * (matrixEnergy / 100.0);
    }

    /**
     * Aberrált sávos sugár – Fourier-összeg, szög-függő harmónikusok.
     * A burok képes az átmérőjét korlátlanul dinamikázni (sávos mód).
     */
    public static double aberratedBase(double thetaDeg, int step, double matrixEnergy,
                                       double retAngleDeg) {
        double theta = Math.toRadians(thetaDeg);
        double energy = matrixEnergy / 100.0;
        double base = 0.40
            + 0.15 * Math.sin(theta * 2 + step * 0.06) * energy
            + 0.08 * Math.cos(theta * 3 - step * 0.04)
            + 0.05 * Math.sin(theta * 5 + step * 0.11) * (1 - energy)
            + 0.03 * Math.cos(theta * 7 - Math.toRadians(retAngleDeg));
        return Math.max(0.05, base);
    }

    // ── Mutató rezgés – hang referencia ────────────────────────────────────--

    public record VibrationResult(double freqHz, double ampNorm,
                                  double arcM, double wavelengthM, String note) {}

    /**
     * Mutató rezgés kiszámítása hangsebesség alapján.
     * arcLength = ARM × |Δθ_rad|
     * freqHz = SOUND / (2 × arcLength)   ha arcLength > 0
     */
    public static VibrationResult calcPointerVibration(double prevAngleDeg, double currAngleDeg) {
        double deltaRad = Math.toRadians(Math.abs(currAngleDeg - prevAngleDeg));
        if (deltaRad > Math.PI) deltaRad = 2 * Math.PI - deltaRad;

        double arcM = ARM * deltaRad;
        if (arcM < 1e-9) {
            return new VibrationResult(0, 0, 0, 0, "-");
        }

        double freqHz      = SOUND / (2.0 * arcM);
        double wavelengthM = SOUND / freqHz;
        double ampNorm     = Math.min(1.0, arcM / (ARM * Math.PI));
        String note        = freqToNote(freqHz);
        return new VibrationResult(freqHz, ampNorm, arcM, wavelengthM, note);
    }

    private static final String[] NOTE_NAMES = {"C","C#","D","D#","E","F","F#","G","G#","A","A#","B"};
    private static String freqToNote(double hz) {
        if (hz <= 0 || hz > 20000) return "-";
        int n   = (int) Math.round(12.0 * Math.log(hz / 440.0) / Math.log(2)) + 69;
        int oct = n / 12 - 1;
        return NOTE_NAMES[((n % 12) + 12) % 12] + oct;
    }

    // ── Kardinális eloszlás ──────────────────────────────────────────────────

    public record Distribution(int north, int east, int south, int west) {
        public String dominant() {
            int mx = Math.max(Math.max(north, east), Math.max(south, west));
            if (north == mx) return "north";
            if (east  == mx) return "east";
            if (south == mx) return "south";
            return "west";
        }
    }

    /** Reticle szög → É/K/D/Ny eloszlás százalékban */
    public static Distribution calcDistribution(double retAngle, double baseR,
                                                double tension, double resonance, int step) {
        double[] angles = {0, 90, 180, 270};
        double[] weights = new double[4];
        double total = 0;
        for (int i = 0; i < 4; i++) {
            double diff = Math.toRadians(normalizeAngle(retAngle - angles[i]));
            double r    = getBalloonRadius(angles[i], baseR, retAngle, tension, resonance, step);
            weights[i]  = Math.max(0, Math.cos(diff)) * r * (1 + resonance * 0.3);
            total      += weights[i];
        }
        if (total == 0) return new Distribution(25, 25, 25, 25);
        int n = (int) Math.round(weights[0] / total * 100);
        int e = (int) Math.round(weights[1] / total * 100);
        int s = (int) Math.round(weights[2] / total * 100);
        int w = 100 - n - e - s;
        return new Distribution(n, e, s, w);
    }

    // ── Segéd ────────────────────────────────────────────────────────────────

    private static double frac(double v) {
        return v - Math.floor(v);
    }

    public static double normalizeAngle(double a) {
        a = a % 360;
        return a < 0 ? a + 360 : a;
    }
}
